AP-protocol.md

```diff
@@ -106,54 +106,73 @@ When the network is first initialized, nodes only know who their own neighbors a
+
+The node then swaps `prev_hop` in the received `FloodRequest` with itself and proceeds to send it to all of its neighbors except the one it just received it from.
+
+The node then waits for replies, and notes down every node from which it receives a `FloodProcessing` response as **pending** in the current flood request.
 
-- If the flood request was not received earlier, the node forwards the updated packet to its neighbors (except the one it received the flood request from) decreasing the TTL by 1, otherwise set the TTL to 0.
-- If the TTL of the message is 0, build a `FloodResponse` and send it along the same path back to the initiator.
+In case a neighbor crashes (AKA a notification from the controller to drop that channel is received), that node is marked as **failed**
 
+If the node is seeing a `FloodRequest` with the same flood ID it has already seen, it MUST ignore the request.
+
+### **Replies, merging and termination condition**
+
+When a **pending** neighbor returns a `FloodData` message, the message is temporarily saved and the node is marked as **done**. When all neighbors are **done** or **failed**, or if there are none (in the case of a node connected to only one neighbor, from which it has received the request) the Node creates its own `FloodData`:
 ```rust
-struct FloodResponse {
-	flood_id: u64,
-	source_routing_header: SourceRoutingHeader,
-	path_trace: Vec<(NodeId, NodeType)>
+pub struct FloodData {
+	// The original flood ID
+    pub flood_id: u64,
+	// Its own node ID
+	pub from:  NodeId
+	// Its own node ID and type
+    pub node_list: HashSet<(NodeId, NodeType)>,
+	// See below for explaination
+    pub connections: HashSet<(NodeId, NodeId)>,
+	// The Node ID of one of the nodes marked as "failed", otherwise None
+    pub error_at: Option<NodeId>,
+	// Kept at "false" 
+    pub is_broadcast: bool 
 }
 ```
+The `connections` field gets populated with the neighbors, creating a tuple of two node IDs, one being the current and one being the neighbor, which MUST be in ascending order: (n1,n2) where n1 < n2 always.
+The node then proceeds to merge all obtained responses, and MUST send the final response to the previous hop it had stored, changing the `from` field to its own ID.
 
-### **Recording Topology Information**
-
-For every flood response or acknowledgment the initiator receives, it updates its understanding of the graph:
-
-- If the node receives a flood response with a **path trace**, it records the paths between nodes. The initiator learns not only the immediate neighbors but also the connections between nodes further out.
-- Over time, as the query continues to flood, the initiator accumulates more information and can eventually reconstruct the entire graph's topology.
+### **Special considerations for Requests containing a "to" field**
```

Comment:
I do not believe that this solution is more efficient than the general solution. In fact, the whole network will still have to be explored to wait for all the neighbors to be done or failed

Reply:
Perhaps one could change the algorithm in case the `to` field is set: a drone could return the `FloodResponse` as soon as it gets a non-empty one, without waiting for all `FloodProcessing` to be resolved.

Reply:
I added an extra provision in response to your comment in the latest commit.